### ISSUE TITLE
Minor issues with PEP 634 (pattern matching)

### ISSUE THREAD

**Issue by gvanrossum** [microsoft/pyright#1934]
Here are a few things I encountered with pylance in strict mode. (Possibly these could be pylance issues.)
```
match x:
    case tuple(t):
        print(len(t))
```
This flags `t` as partially unknown and complains about the arg to `len()` too. But adding a type declaration for `t` would be ugly and perhaps incorrect (it'd have to go before the match). Since there's no good solution, maybe in this case the type should not be flagged? It's possible that this is purely a pylance issue or pylance is using an older version of pyright.

```
def add_float(re: float, im: float): ...

match x:
    case complex(real=re, imag=im):
        foo(re, im)  # Error
```
This I get with pyright 1.1.145 as well. The error is
```
Argument of type "property" cannot be assigned to parameter "value" of type "float" in function "add_float"
    "property" is incompatible with "float" (reportGeneralTypeIssues)
```

**Comment by erictraut** [comment]
I haven't changed anything in pyright's pattern matching code in over a month, so pylance should be up to date here.

In the first example, what is the type of `x` prior to the match statement? It works fine if it's a tuple or a union of tuples.

```python
def func2(x: Tuple[int, ...] | Tuple[str, ...]):
    match x:
        case tuple(t):
            reveal_type(t) # Tuple[int, ...] | Tuple[str, ...]
            print(len(t))
```

I'm guessing that you've defined it as something else, like `Sequence[int]`?

For the second sample, I don't think I have enough context to understand what this code is meant to do. Could you provide a bit more detail? Is `foo` meant to be `add_float`? If I replace `foo` with `add_float`, I don't see that error message. Ah, I think I figured it out. If `x` is typed as `complex` or a union that contains `complex`, then I see this error. It's because the type checker isn't properly handling properties in this context.

**Comment by gvanrossum** [comment]
Sure. You can see the full code here: https://github.com/faster-cpython/tools/blob/af5b6b39b6adc8a2f1fed6bd515df20bed080da7/pyco/pyco.py#L121

In the `tuple` example, `x` is `object`, and there are a bunch of other cases. Your question made me find a way to make the errors go away, but it's pretty verbose: declare the argument as a union of all the types captured by the cases: I change the argument type to
```
value: None | complex | bytes | str | tuple[object]
```

For the `complex`, here's a self-contained example:
```
def f(a: float):
    print(a)

def foo(a: object):
    match a:
        case complex(real=re, imag=im):
            f(re)
            f(im)
```
This wavy-underlines the `re` and `im` arguments to the f() calls on the last two lines. Running pyright standalone gives:
```
PS C:\Users\gvanrossum\tools> pyright t.py --pythonversion 3.10 
No configuration file found.
No pyproject.toml file found.
stubPath C:\Users\gvanrossum\tools\typings is not a valid directory.
Assuming Python platform Windows
Searching for source files
Found 1 source file
C:\Users\gvanrossum\tools\t.py
  C:\Users\gvanrossum\tools\t.py:7:15 - error: Argument of type "property" cannot be assigned to parameter "a" of type "float" in function "f"
    "property" is incompatible with "float" (reportGeneralTypeIssues)
  C:\Users\gvanrossum\tools\t.py:8:15 - error: Argument of type "property" cannot be assigned to parameter "a" of type "float" in function "f" 
    "property" is incompatible with "float" (reportGeneralTypeIssues)
2 errors, 0 warnings, 0 infos
Completed in 0.617sec
PS C:\Users\gvanrossum\tools> 
```

To me it looks as if it takes the definition of `complex`, which I presume (without looking) must be something like this:
```
class complex:
    @property
    def real(self) -> float: ...
    @property
    def imag(self) -> float: ...
```
and somehow thinks that the match statement puts `complex.real` and `complex.imag` in the variables rather than `complex().real` and `complex().imag`.

**Comment by erictraut** [comment]
For the tuple example, I think the current behavior is correct. It's consistent with the way pyright generates "Unknown" in other cases, and I think it would be inappropriate to infer a `Tuple[Any, ...]` type in this case.

The reason that "Unknown" exists is to give the programmer visibility into a "blind spot" in strict mode in cases where the type checker is unable to infer a known type. We never want to implicitly infer an `Any` type.

Consider the example:
```python
def func(x: object):
    match x:
        case tuple(t):
            return t[0] ** 2 # This will generate runtime errors if t[0] doesn't support __pow__
```

I mentioned earlier that class pattern matching would interact poorly with static type checking. This is a good example.

If you're interested in strict type checking, then you'll need to avoid certain uses of class pattern matching like this. Or you can suppress errors on a line-by-line or file-by-file basis as you see fit.

The other issue is definitely a bug. As you suspected, the type checker was not applying descriptor logic, so property or descriptor attributes that were referenced in pattern arguments did not receive the correct type. The fix for this will be in the next release of pyright and pylance.

**Comment by gvanrossum** [comment]
I was afraid you'd say that (about the tuple example), and I agree it's consistent. It's just annoying for strictness-completionists like myself :-) -- I really don't like having to use any of the ways to suppress the wiggly underline.

I should probably figure out how to disable this particular strictness flag.

If this becomes common in the future we may have to consider augmenting the syntax for patterns to allow inserting type annotations. In this example it's easy enough, you could write `case tuple(t: tuple[object]):`, although it's more awkward if there are keyword parameters in the pattern.

**Comment by gvanrossum** [comment]
And thanks for confirming/fixing the bug for the other example.

**Comment by erictraut** [comment]
This is included in pyright 1.1.146, which I just published. It will also be included in the next release of pylance.

**Comment by gvanrossum** [comment]
Confirmed. Thanks!
